vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -226,211 +270,195 @@ facebook::jsi::Runtime::PointerValue *ChakraRuntime::clonePropNameID(
-  return Common::Unicode::Utf16ToUtf8(GetPropertyName(GetChakraObjectRef(id)));
+  return Common::Unicode::Utf16ToUtf8(GetPropertyNameFromId(GetJsRef(id)));
 }
 
 bool ChakraRuntime::compare(const facebook::jsi::PropNameID &lhs, const facebook::jsi::PropNameID &rhs) {
-  return CompareJsPropertyIds(GetChakraObjectRef(lhs), GetChakraObjectRef(rhs));
+  return GetJsRef(lhs) == GetJsRef(rhs);
 }
 
 std::string ChakraRuntime::symbolToString(const facebook::jsi::Symbol &s) {
-  return ToStdString(ToJsString(GetChakraObjectRef(s)));
+  JsValueRef symbol = GetJsRef(s);
+  JsValueRef symbolCtor = GetProperty(GetGlobalObject(), m_propertyId.Symbol);
+  JsValueRef symbolPrototype = GetProperty(symbolCtor, m_propertyId.prototype);
+  JsValueRef symbolToString = GetProperty(symbolPrototype, m_propertyId.toString);
+  JsValueRef jsString = CallFunction(symbolToString, {symbol});
+  return StringToStdString(jsString);
 }
 
 facebook::jsi::String ChakraRuntime::createStringFromAscii(const char *str, size_t length) {
-  // Unfortunately due to the typing used by JSI and Chakra, we have to do a
-  // double reinterpret cast here.
-  return createStringFromUtf8(Common::Utilities::CheckedReinterpretCast<const uint8_t *>(str), length);
+  return MakePointer<facebook::jsi::String>(PointerToString({str, length}));
 }
 
 facebook::jsi::String ChakraRuntime::createStringFromUtf8(const uint8_t *str, size_t length) {
-  return MakePointer<facebook::jsi::String>(
-      ToJsString(std::string_view{Common::Utilities::CheckedReinterpretCast<const char *>(str), length}));
+  return MakePointer<facebook::jsi::String>(PointerToString({reinterpret_cast<char const *>(str), length}));
 }
 
 std::string ChakraRuntime::utf8(const facebook::jsi::String &str) {
-  return ToStdString(GetChakraObjectRef(str));
+  return StringToStdString(GetJsRef(str));
 }
 
 facebook::jsi::Object ChakraRuntime::createObject() {
-  JsValueRef obj;
-  VerifyJsErrorElseThrow(JsCreateObject(&obj));
-  return MakePointer<facebook::jsi::Object>(obj);
+  return MakePointer<facebook::jsi::Object>(CreateObject());
 }
 
 facebook::jsi::Object ChakraRuntime::createObject(std::shared_ptr<facebook::jsi::HostObject> hostObject) {
-  facebook::jsi::Function jsProxyConstructor =
-      global().getPropertyAsFunction(*this, g_proxyConstructorBootstrapFuncName);
-
-  return jsProxyConstructor
-      .call(*this, MakePointer<facebook::jsi::Object>(ToJsObject(hostObject)), createHostObjectProxyHandler())
-      .asObject(*this);
+  // The hostObjectHolder keeps the hostObject as external data.
+  // Then the hostObjectHolder is wrapped up by a Proxy object to provide access to hostObject's
+  // get, set, and getPropertyNames methods.
+  // There is a special symbol property ID 'hostObjectSymbol' to access the hostObjectWrapper from the Proxy.
+  JsValueRef hostObjectHolder =
+      CreateExternalObject(std::make_unique<std::shared_ptr<facebook::jsi::HostObject>>(std::move(hostObject)));
```

Comment:
>make_unique<std::shared_ptr [](start = 32, length = 27)

is this right? to make a unique_ptr of a shared_ptr?

Reply:
Yes, we effectively allocate a new object that only contains the shared_ptr because we need to give void* to the ChakraAPI. Using IUnknown would be better than shared_ptr because we could just cast it instead. We use unique_ptr instead of operator new to have better lifetime management for the passed value in case of exceptions.

---
In reply to: [470467183](https://github.com/microsoft/react-native-windows/pull/5710#discussion_r470467183) [](ancestors = 470467183)